src/streamed_orderbook.ts

```diff
@@ -0,0 +1,267 @@
+import { BatchExchange } from '@gnosis.pm/dex-contracts/build/types/BatchExchange'
+
+const STARTING_BLOCK = 9340147
+const BLOCK_PAGE_SIZE = 10000
+const BATCH_DURATION = 300
+
+interface AccountsState {
+  accounts: { [key: string]: AccountState },
+  tokens: { [key: string]: string },
+  prices: { [key: string]: bigint },
+}
+
+interface AccountState {
+  balances: { [key: string]: bigint },
+  withdrawals: { [key: string]: bigint },
+  orders: (OrderState | null)[],
+}
+
+interface OrderState {
+  buyToken: string,
+  sellToken: string,
+  validFrom: number,
+  validUntil: number | null,
+  priceNumerator: bigint,
+  priceDenominator: bigint,
+  remainingAmount: bigint,
+}
+
+class StreamedOrderbook {
+  private initialized: boolean = false;
+  private blockNumber: number = 0;
+  private blockBatches: { [key: number]: number } = {};
+  private pendingEvents: any[] = [];
+  private state: AccountsState = {
+    accounts: {},
+    tokens: {},
+    prices: {},
+  };
+
+  constructor(
+    private readonly web3: Web3,
+    private readonly contract: BatchExchange,
+  ) {}
+
+  async init() {
+    if (this.initialized) {
+      return
+    }
+    this.initialized = true
+
+    this.blockNumber = await this.web3.eth.getBlockNumber()
+    for (
+      let fromBlock = STARTING_BLOCK;
+      fromBlock < this.blockNumber;
+      fromBlock += BLOCK_PAGE_SIZE
+    ) {
+      const toBlock = Math.min(fromBlock + BLOCK_PAGE_SIZE - 1, this.blockNumber)
+      console.error(`fetching page ${fromBlock}-${toBlock}`)
+
+      const events = await this.contract.getPastEvents(
+        'allEvents',
+        { fromBlock, toBlock },
+      )
+
+      this.pendingEvents.push(...events)
+      await this.applyAllEventsUntilNextBatch()
+    }
+  }
+
+  dump() {
+    for (const user in this.state.accounts) {
+      console.log(user, this.state.accounts[user])
+    }
+  }
+
+  private async applyAllEventsUntilNextBatch() {
+    if (this.pendingEvents.length === 0) {
+      return
+    }
+
+    const pending = this.pendingEvents.length
+    const nextBatch = await this.eventBatch(this.pendingEvents[pending - 1])
+    for (let i = pending - 2; i >= 0; i--) {
+      const batch = await this.eventBatch(this.pendingEvents[i])
+      if (nextBatch > batch) {
+        this.applyEvents(this.pendingEvents.splice(0, i + 1))
+        return
+      }
+    }
+
+    return false
+  }
+
+  private applyEvents(events: EventData[]) {
+    this.markReverted(events)
+    for (const ev of events) {
+      const data = ev.returnValues
+      switch (ev.event) {
+        case 'Deposit':
+          this.updateBalance(
+            data.user,
+            data.token,
+            amount => amount + BigInt(data.amount),
+          )
+          break
+        case 'OrderCancellation':
+          this.account(data.owner).orders[parseInt(data.id)]!.validUntil = null
+          break
+        case 'OrderDeletion':
+          this.account(data.owner).orders[parseInt(data.id)] = null
```

Comment:
To be closer to the smart contract, would it make sense to actually use 0s here?

Reply:
Sure, I'm OK with that.